third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,1161 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! AST Simplifier
+//!
+//! Simplify the AST before conversion to bytecode.
+//! - flow-insensitive constant propagation and folding
+//! - simple expression simplification
+//!
+//! Preconditions:
+//! - have previously run `check_for_unused_vars_and_params` pass to warn about unused vars,
+//!   which may be eliminated here.
+//!
+//! More details:
+//! - Do flow-insensitive constant propagation:
+//!   - identify "possibly modified" symbols which may have more than one value
+//!   - for safe symbols whose value is a constant value, propagate
+//!     the value to the use site to enable simplifying code:
+//!     - inline a constant value
+//!     - flag uses of uninitialized and unassigned variables
+//! - Implement ExpRewriterFunctions to allow bottom-up replacement
+//!   of some complex expressions by "simpler" ones:
+//!   - Constant folding of operations with constant parameters
+//!   - Eliminate unused variables (with a warning)
+//!   - Eliminate used variables whose uses are all eliminated by
+//!     constant folding
+//!   - Eliminate unused value expressions which are side-effect-free.
+//!   - Unwrap trivial compound expressions:
+//!     - a Sequence of 1 expression
+//!     - a Block with no variable binding
+//!   - Simple call rewriting: (one example)
+//!     - eliminate cast to same type as parameter
+//!
+//! - Optionally do some simplifications that may eliminate dead
+//!   code and hide some warnings:
+//!     - eliminate side-effect-free expressions with ignored value
+//!       in a `Sequence` instruction.
+//!     - eliminate unused variable assignments in a `let` statement,
+//!       and unassigned values expressions from `let` RHS which are
+//!       side-effect-free.
+//!     - use constant folding on if predicates to eliminate dead
+//!       then or else branches (currently disabled by local constant,
+//!       as it may eliminate some useful code diagnostics).
```

Comment:
We can at least eliminate an unused empty branch here if the condition is constant.  Actually, I guess I could do that even if `eliminate_code` is false but since it's disabled by default anyway perhaps it is ok to checkin and we can discuss further once it's in and I can stop rebasing.